proposals/0000-app-setup.md

```diff
@@ -0,0 +1,184 @@
+
+```js
+{
+    async resolveId(id, importer, options) {
+        if (id.startsWith('virtual:') && id.endsWith('/app')) {
+            const rendererName = id.slice('virtual:'.length, '/app'.length * -1);
+            const match = config._ctx.renderers.find(({ name }) => name === rendererName);
+            if (match && match.appEntrypoint) {
+                // Resolve `appEntrypoint` relative to configured root
+                const app = await this.resolve(match.appEntrypoint, config.root, { ...options, skipSelf: true });
+                return app;
+            }
+            // Fallback to a shared `noop`
+            return `\0virtual:noop`;
+        }
+    },
+    load(id) {
+        if (id == `\0virtual:noop`) {
+            return `export default () => {}`;
+        }
+    }
+}
+```
+
+## `appEntrypoint` in client and server entrypoints
+
+For each supported renderer, both the client and server entrypoints should import and use this virtual module.
+
+### Vue
+
+```js
+import setup from 'virtual:@astrojs/vue/app'
+
+async function renderToStaticMarkup(Component, props, children) {
+  // ...
+  const app = createSSRApp({ render: () => h(Component, props, children) })
+  await setup({ app })
+  // ...
+}
+```
+
+### React and Preact
+
+The `setup` function should return a component that will be used as a global provider for every island.
+
+> **Warning**:
+> This RFC explicity does **NOT** concern itself with **sharing state between islands**.
+> The component returned from this function will **NOT** be a singleton and state will not automatically synchronize between islands.
+
+```js
+import setup from 'virtual:@astrojs/react/app'
+
+async function renderToStaticMarkup(Component, props, children) {
+  // ...
+  const Provider = await setup()
+  const html = h(Provider, {}, h(Component, props, children))
+  // ...
+}
+```
+
+# Drawbacks
+
+This is a relatively light feature to implement, but it does come with some gotchas:
+
+- This proposal exposes lower-level control to users. Fine for advanced use cases, but makes it easier to introduce performance footguns if not used carefully.
+- This proposal introduces a new pattern that, while modelled after similar patterns in other frameworks, has a unique design due to Astro's constraints
+- This proposal **DOES NOT** solve a common problem for sharing state via `Provider`s. It's extremely likely that people will expect this feature to do that because you can setup a Provider. However, Providers in this proposal **ARE NOT** stateful between islands—updates to one Provider instance will not automatically propogate to other Provider instances unless specifically architected to do so by the user.
```

Comment:
I understand the desire to avoid feature-bloat, but presumably you are thinking about this. Shared context in React is one very desirable feature. We can skip that for now, but it would be good if you have an idea of what that might look like so that this feature can be a stepping stone and that is not a completely different feature, if that makes sense.

Reply:
Yeah that totally makes sense. This proposal is certainly not blocking an implementation of shared context in the future, since a prototype of that is what I extracted this proposal from. I was hoping to avoid this conversation being focused on shared state because there's still a lot to figure out there, but users asking to hook into `vue.use` comes up almost every day.

Reply:
I would even be open to not implementing this `appEntrypoint` proposal for React/Preact until shared state lands. It's possible this might just be a footgun until Providers can be singletons that only run once.

Reply:
I think holding off on React makes sense. I haven't seen as much of a demand for the "pyramid of providers" pattern as there has been for Vue .use() so far.

Reply:
It would be quite neat if there was a way to specify setup code that only gets loaded on server side or client side (like in the game engine melonJS, it would be cool to render the game to a static image on the server side, render that, then on the client side replace it with the actual game)